src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java

```diff
@@ -2,37 +2,64 @@
 import com.iridium.iridiumskyblock.IridiumSkyblock;
-import com.iridium.iridiumskyblock.api.IridiumSkyblockAPI;
+import com.iridium.iridiumskyblock.api.IslandEnterEvent;
+import com.iridium.iridiumskyblock.api.IslandLeaveEvent;
 import com.iridium.iridiumskyblock.database.Island;
 import com.iridium.iridiumskyblock.database.IslandBooster;
 import com.iridium.iridiumskyblock.database.User;
+import org.bukkit.Bukkit;
 import org.bukkit.GameMode;
 import org.bukkit.entity.Player;
 import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
 import org.bukkit.event.Listener;
 import org.bukkit.event.player.PlayerMoveEvent;
 
 import java.util.Optional;
 
 public class PlayerMoveListener implements Listener {
 
-    @EventHandler
+    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
     public void onPlayerMove(PlayerMoveEvent event) {
         Player player = event.getPlayer();
         User user = IridiumSkyblock.getInstance().getUserManager().getUser(player);
-        if (event.getTo().getBlockX() != event.getFrom().getBlockX() || event.getTo().getBlockZ() != event.getFrom().getBlockZ()) {
+        if (event.getTo() != null && (event.getTo().getBlockX() != event.getFrom().getBlockX() || event.getTo().getBlockZ() != event.getFrom().getBlockZ())) {
             if (user.getTeleportingTask() != null) {
                 user.getTeleportingTask().cancel();
+
                 user.setTeleportingTask(null);
                 player.sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().teleportCanceled
                         .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix))
                 );
             }
+            Optional<Island> fromIsland = IridiumSkyblock.getInstance().getIslandManager().getIslandViaLocation(event.getFrom());
+            Optional<Island> toIsland = IridiumSkyblock.getInstance().getIslandManager().getIslandViaLocation(event.getTo());
+            if ((!fromIsland.isPresent() && toIsland.isPresent()) || (fromIsland.isPresent() && toIsland.isPresent() && !fromIsland.get().equals(toIsland.get()))) {
+                if (IridiumSkyblock.getInstance().getIslandManager().isBannedOnIsland(toIsland.get(), user)) {
+                    event.getPlayer().sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().youHaveBeenBanned.replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix).replace("%owner%", toIsland.get().getOwner().getName())));
+                    event.setCancelled(true);
+                    return;
+                }
 
+                Bukkit.getPluginManager().callEvent(new IslandEnterEvent(toIsland.get(), user));
+                toIsland.get().getPlayersOnIsland().add(player);
+                //TODO add enter island message if we want, i'll remove below message  when pr finished
+                Bukkit.broadcastMessage("§aWelcome to " + toIsland.get().getOwner().getName() + "'s island");
```

Comment:
If this pr isnt ready to be merged you know you can create draft pr's right?

Reply:
no i just waiting comment from u, we want enter/leave message or no, if no i'll remove these lines

Reply:
also im not known whats draft pr lol

Reply:
https://imgur.com/a/rb4IYtG

Reply:
It's like a normal PR but you can't merge it and it's obvious that it's not ready for usage

Reply:
nvm just say yes or no, we want messages?

Reply:
No, don't do it 👀

Reply:
I don't really care

Reply:
Im 99% sure that was a suggestion, I say add it but make it so if its blank it doesnt send